src/backend/joanie/tests/core/debug/test_debug_urls_template_preview.py

```diff
@@ -0,0 +1,132 @@
+"""
+Test suite for Debug app urls for preview template (certificate, degree, invoice and
+contract definition)
+"""
+from http import HTTPStatus
+
+from django.core.cache import cache
+from django.test import TestCase
+from django.test.utils import override_settings
+
+from joanie.tests.testing_utils import reload_urlconf
+
+
+@override_settings(DEBUG=True)
+class DebugUrlsTemplatePreviewTestCase(TestCase):
+    """
+    Test case for debug urls routes to preview template
+    (certificate, degree, invoice and contract definition)
+    """
+
+    def setUp(self):
+        super().setUp()
+        # Reset the cache to always reach the site route.
+        cache.clear()
+        # Force URLs reload to take DEBUG into account
+        reload_urlconf()
+
+    def test_debug_urls_template_preview_contract_definition_when_debug_is_set_to_true(
+        self,
+    ):
+        """
+        When `DEBUG` is set to `True`, the developer can access the view of the contract
+        definition to preview the template.
+        """
+        response = self.client.get(path="/__debug__/pdf-templates/contract")
+
+        self.assertEqual(response.status_code, HTTPStatus.OK)
+
+    def test_debug_urls_template_preview_degree_when_debug_is_set_to_true(
+        self,
+    ):
+        """
+        When `DEBUG` is set to `True`, the developer can access the view of the degree
+        to preview the template.
+        """
+        response = self.client.get(path="/__debug__/pdf-templates/degree")
+
+        self.assertEqual(response.status_code, HTTPStatus.OK)
+
+    def test_debug_urls_template_preview_certificate_when_debug_is_set_to_true(
+        self,
+    ):
+        """
+        When `DEBUG` is set to `True`, the developer can access the view of the certificate
+        to preview the template
+        """
+        response = self.client.get(path="/__debug__/pdf-templates/certificate")
+
+        self.assertEqual(response.status_code, HTTPStatus.OK)
+
+    def test_debug_urls_template_preview_invoice_when_debug_is_set_to_true(
+        self,
+    ):
+        """
+        When `DEBUG` is set to `True`, the developer can access the view of an invoice
+        to preview the template
+        """
+        response = self.client.get(path="/__debug__/pdf-templates/invoice")
+
+        self.assertEqual(response.status_code, HTTPStatus.OK)
+
+    @override_settings(DEBUG=False)
+    def test_debug_urls_template_preview_contract_definition_when_debug_is_set_to_false(
+        self,
+    ):
+        """
+        When `DEBUG` is set to `False`, a user cannot access the view of the contract
+        definition to preview the template because urlpattern does not exist.
+        """
+        # Reset the cache to always reach the site route.
+        cache.clear()
+        # Force URLs reload to take DEBUG into account
+        reload_urlconf()
```

Comment:
As this is in the setup, is it still needed to keep it ?

Reply:
If I don't declare it again within the method, the test will fail by returning an HTTP.OK status code. 
* without setUp() method : every test linked to DEBUG is True will fail.
* without declaring 'cache.clear' and reload_urlconf() within the method : every test linked to DEBUG is False will fail.
* using override_settings as a context manager (with the 'with' statement):  every test linked to DEBUG is False will fail.

Reply:
For more clarity, we have decided with @jbpenrath to create two distinct classes to test when DEBUG is **enabled**, and when it is **disabled**.